src/kafka-extensions/registry.ts

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * Cluster provider is used to:
+ *
+ *  - collect clusters (eg: create a cluster from a wizard, import clusters from a repository, ...)
+ * and add thento the Kafka Explorer.
+ *  - create a Kafka client from a complex process (eg : use SSO to connect to the cluster)
+ *
+ * Implementing a cluster provider in custom vscode extension is done in 2 steps:
+ *
+ *  - define the cluster provider (id, name) in the package.json in the contributes/kafka/clusterProviders section.
+ *  - return the cluster provider processor (See ClusterProviderProcessor) to use in the activate() of the extension.
+ *
+ */
+export class ClusterProvider {
+
+    private processor: ClusterProviderProcessor | undefined;
+
+    constructor(private definition: ClusterProviderDefinition, private extensionId: string) {
+
+    }
+
+    /**
+     * Returns the cluster provider id.
+     */
+    public get id(): string {
+        return this.definition.id;
+    }
+
+    /**
+     * Returns the cluster provider name.
+     */
+    public get name(): string {
+        return this.definition.name || this.definition.id;
+    }
+
+    /**
+     * Returns the clusters managed by the provider which must be added to the kafka explorer.
+     *
+     * @param clusterSettings the cluster settings.
+     */
+    async collectClusters(clusterSettings: ClusterSettings): Promise<Cluster[] | undefined> {
+        const processor = await this.getProcessor();
+        return processor.collectClusters(clusterSettings);
+    }
+
+    /**
+     * Create the Kafka JS client instance from the given connection options.
+     *
+     * @param connectionOptions the connection options.
+     */
+    async createKafkaConfig(connectionOptions: ConnectionOptions): Promise<KafkaConfig | undefined> {
+        const processor = await this.getProcessor();
+        if (processor.createKafkaConfig) {
+            return processor.createKafkaConfig(connectionOptions);
+        }
+    }
+
+    private async getProcessor(): Promise<ClusterProviderProcessor> {
+        if (this.processor) {
+            return this.processor;
+        }
+        // The cluster provider processor is not already loaded, try to activate the owner extension.
+        // The return of the extension activate() method must return the cluster provider processor.
+        const extension = vscode.extensions.getExtension(this.extensionId);
+        if (!extension) {
+            throw new Error(`Error while getting cluster provider processor. Extension ${this.extensionId} is not available.`);
+        }
+
+        // Wait for extension is activated to get the processor
+        const result = await extension.activate();
+        if (!result) {
+            throw new Error(`Error while getting cluster provider processor. Extension ${this.extensionId}.activate() should return 'ClusterProviderProcessor' or 'ClusterProviderProcessorProvider'.`);
+        }
+        if ('getClusterProviderProcessor' in result) {

```

Comment:
it'll be simpler if we only expect the  result to be an instance of KafkaExtensionParticipant

Reply:
I have tried to do that, have some piece of code which could check that activate method result is an instanceof KafkaExtensionParticipant.

Reply:
I have tried to improve this instanceof. Hope it's the fix that you expected.